pkg/util/iptree/iptree.go

```diff
@@ -0,0 +1,592 @@
+	return out
+}
+
+// WalkPrefix is used to walk the tree under a prefix
+func (t *Tree[T]) WalkPrefix(prefix netip.Prefix, fn WalkFn[T]) {
+	n := t.rootV4
+	if prefix.Addr().Is6() {
+		n = t.rootV6
+	}
+	bitPosition := 0
+	// mask the address for sanity
+	address := prefix.Masked().Addr()
+	// we can't check longer than the request mask
+	mask := prefix.Bits()
+	// walk the network bits of the prefix
+	for bitPosition < mask {
+		// Look for a child checking the bit position after the mask
+		n = n.child[getBitFromAddr(address, bitPosition+1)]
+		if n == nil {
+			return
+		}
+		// check we are in the right branch comparing the suffixes
+		if !n.prefix.Contains(address) {
+			break
+		}
+		// update the new bit position with the new node mask
+		bitPosition = n.prefix.Bits()
+	}
+	recursiveWalk[T](n, fn)
+
+}
+
+// WalkPath is used to walk the tree, but only visiting nodes
+// from the root down to a given public. Where WalkPrefix walks
+// all the entries *under* the given prefix, this walks the
+// entries *above* the given prefix.
+func (t *Tree[T]) WalkPath(path netip.Prefix, fn WalkFn[T]) {
+	n := t.rootV4
+	if path.Addr().Is6() {
+		n = t.rootV6
+	}
+	bitPosition := 0
+	// mask the address for sanity
+	address := path.Masked().Addr()
+	// we can't check longer than the request mask
+	mask := path.Bits()
+	// walk the network bits of the prefix
+	for bitPosition < mask {
+		// Visit the public values if any
+		if n.isPublic() && fn(n.prefix, n.val) {
+			return
+		}
+		// Look for a child checking the bit position after the mask
+		n = n.child[getBitFromAddr(address, bitPosition+1)]
+		if n == nil {
+			return
+		}
+		// check we are in the right branch comparing the suffixes
+		if !n.prefix.Contains(address) {
+			return
+		}
+		// update the new bit position with the new node mask
+		bitPosition = n.prefix.Bits()
+	}
+	// check if this node is a public node and contains o prefix
+	if n != nil && n.isPublic() && n.prefix == path {
+		fn(n.prefix, n.val)
+	}
+}
+
+// TopLevelPrefixes is used to return a map with all the Top Level prefixes
+// from the corresponding IP family and its values
+func (t *Tree[T]) TopLevelPrefixes(isIPv6 bool) map[string]T {
+	if isIPv6 {
+		return t.topLevelPrefixes(t.rootV6)
+	}
+	return t.topLevelPrefixes(t.rootV4)
+}
+
+// topLevelPrefixes is used to return a map with all the Top Level prefixes and its values
+func (t *Tree[T]) topLevelPrefixes(root *node[T]) map[string]T {
+	result := map[string]T{}
+	queue := []*node[T]{root}
+
+	for len(queue) > 0 {
+		n := queue[0]
+		queue = queue[1:]
+		// store and continue, only interested on the top level prefixes
+		if n.isPublic() {
+			result[n.prefix.String()] = n.val
+			continue
+		}
+		if n.child[0] != nil {
+			queue = append(queue, n.child[0])
+		}
+		if n.child[1] != nil {
+			queue = append(queue, n.child[1])
+		}
+	}
+	return result
+}
+
+// assume starts at 0 from the MSB: 0.1.2......31
+// return 0 or 1
+func getBitFromAddr(ip netip.Addr, pos int) int {
```

Comment:
I think we can simplify this.
1. Normalize `pos` to zero based pos (no need to add one by caller and deducted here).
2. Preflight boundaries: `if(pos < 0 || !(pos < len(slice)))`.
3. Create a mask separately `mask := 1 << (pos % 32)`
4. Get an int by its position and mask it `return Binary.LittleEndian(slice[pos/32:8]) & mask`

Reply:
ipv6 addresses is the problem, so we can get a 4 or 16 bytes array.
To share the same code we calculate first which byte in the array contains the pos we want, and then we mask that byte.